tokio/src/sync/once_cell.rs

```diff
@@ -0,0 +1,313 @@
+    pub fn initialized(&self) -> bool {
+        self.value_set.load(Ordering::Acquire)
+    }
+
+    // SAFETY: safe to call only once self.initialized() is true
+    unsafe fn get_unchecked(&self) -> &T {
+        &*self.value.with(|ptr| (*ptr).as_ptr())
+    }
+
+    // SAFETY: safe to call only once a permit on the semaphore has been
+    // acquired
+    unsafe fn set_value(&self, value: T) {
+        self.value.with_mut(|ptr| (*ptr).as_mut_ptr().write(value));
+        self.value_set.store(true, Ordering::Release);
+        self.semaphore.close();
+    }
+
+    /// Tries to get a reference to the value of the OnceCell.
+    ///
+    /// Returns [`NotInitializedError`] if the value of the OnceCell
+    /// hasn't previously been initialized.
+    ///
+    /// [`NotInitializedError`]: crate::sync::NotInitializedError
+    pub fn get(&self) -> Result<&T, NotInitializedError> {
+        if self.initialized() {
+            Ok(unsafe { self.get_unchecked() })
+        } else {
+            Err(NotInitializedError)
+        }
+    }
+
+    /// Sets the value of the OnceCell to the argument value.
+    ///
+    /// If the value of the OnceCell was already set prior to this call
+    /// or some other set is currently initializing the cell, then
+    /// [`AlreadyInitializedError`] is returned. In order to wait
+    /// for an ongoing initialization to finish, call [`OnceCell::get_or_init`]
+    /// or [`OnceCell::get_or_init_with`] instead.
+    ///
+    /// [`AlreadyInitializedError`]: crate::sync::AlreadyInitializedError
+    /// ['OnceCell::get_or_init`]: crate::sync::OnceCell::get_or_init
+    /// ['OnceCell::get_or_init_with`]: crate::sync::OnceCell::get_or_init_with
+    pub fn set(&self, value: T) -> Result<(), AlreadyInitializedError> {
+        if !self.initialized() {
+            // After acquire().await we have either acquired a permit while self.value
+            // is still uninitialized, or another thread has intialized the value and
+            // closed the semaphore, in which case self.initialized is true and we
+            // don't set the value
+            match self.semaphore.try_acquire() {
+                Ok(_permit) => {
+                    if !self.initialized() {
+                        // SAFETY: There is only one permit on the semaphore, hence only one
+                        // mutable reference is created
+                        unsafe { self.set_value(value) };
+
+                        return Ok(());
+                    } else {
+                        unreachable!(
+                            "acquired the permit after OnceCell value was already initialized."
+                        );
+                    }
+                }
+                _ => {
+                    if !self.initialized() {
+                        panic!(
+                            "couldn't acquire a permit even though OnceCell value is uninitialized."
+                        );
+                    }
```

Comment:
This failure can actually happen if someone has acquired the permit but is currently running the async function. Maybe we should have a test for this case.